lib/rspec/core/memoized_helpers.rb

```diff
@@ -123,14 +123,28 @@ def is_expected
 
       # @private
       def __memoized
-        @__memoized ||= {}
+        @__memoized ||= Memoized.new
+      end
+
+      # @private
+      class Memoized
+        def initialize
+          @memoized = {}
+          @mutex = ReentrantMutex.new
+        end
+
+        def fetch_or_store(key)
+          @mutex.synchronize do
+            @memoized.fetch(key) { @memoized[key] = yield }
+          end
+        end
```

Comment:
Nevermind -- I have no idea what I was thinking but I'm convinced now I was wrong.  Sorry for the false alarm, though.

I do have one suggestion for a potential perf improvement...I think `fetch_or_store` could be:

``` ruby
def fetch_or_store(key)
  @memoized.fetch(key) do
    @mutex.synchronized do
      @memoized.fetch(key) { @memoized[key] = yield }
    end
  end
end
```

That way the synchronization cost is only paid when the `@memoized` doesn't yet have the value.  On future accesses, it short-circuits and just returns the value it already has w/o synchronizing.